File: v4/rust/iprange-livedb/src/bitmap_page.rs
//! Canonical hierarchical-bitmap page layout and scalar access.

use crate::contract::{u32_le, u64_le, PAGE_MAGIC, PAGE_SIZE};
use crate::error::{Error, Result};
use crate::format::page_type;
use crate::mapping::ByteSource;
use crate::page_header;
use crate::slotted_page::{PageEdit, PageSink, HEADER_SIZE};

pub(crate) const BRANCH_TYPE: u8 = page_type::USED_BITMAP_BRANCH;
pub(crate) const LEAF_TYPE: u8 = page_type::USED_BITMAP_LEAF;
pub(crate) const LEAF_WORDS: usize = 500;
pub(crate) const LEAF_BITS: u64 = (LEAF_WORDS * 64) as u64;
pub(crate) const BRANCH_CHILDREN: usize = 256;
pub(crate) const LEAF_END: usize = HEADER_SIZE + LEAF_WORDS * 8;
pub(crate) const BRANCH_END: usize = HEADER_SIZE + 32 + BRANCH_CHILDREN * 4;
pub(crate) const MAX_LEVEL: u16 = 3;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[repr(u32)]
pub(crate) enum Kind {
    Free = 1,
    Feed = 2,
    Membership = 3,
}

impl Kind {
    pub(crate) const fn first_candidate(self) -> u64 {
        match self {
            Self::Membership => 1,
            Self::Free | Self::Feed => 0,
        }
    }
}

#[derive(Clone, Copy)]
pub(crate) struct Header {
    pub(crate) level: u16,
    pub(crate) item_count: usize,
}

#[derive(Clone, Copy)]
pub(crate) enum HeaderProblem {
    Header,
    Born,
    Level,
    Type,
}

pub(crate) fn inspect_header<S: ByteSource>(
    page: S,
    selected_txn: u64,
    kind: Kind,
    expected_level: Option<u16>,
) -> Result<Header> {
    checked_header(page, selected_txn, kind, expected_level)
        .map_err(|_| Error::Corrupt("bitmap page header is invalid"))
}

pub(crate) fn checked_header<S: ByteSource>(
    page: S,
    selected_txn: u64,
    kind: Kind,
    expected_level: Option<u16>,
) -> std::result::Result<Header, HeaderProblem> {
    if let Some(problem) = header_problem(page, selected_txn, kind, expected_level) {
        return Err(problem);
    }
    Ok(Header {
        level: page_header::level(page),
        item_count: page_header::item_count(page),
    })
}

pub(crate) fn header_problem<S: ByteSource>(
    page: S,
    selected_txn: u64,
    kind: Kind,
    expected_level: Option<u16>,
) -> Option<HeaderProblem> {
    let level = page_header::level(page);
    let lower = page_lower(level);
    if !page_header::common_valid(page)
        || page_header::lower(page) != lower
        || page_header::upper(page) != PAGE_SIZE
    {
        return Some(HeaderProblem::Header);
    }
    if !page_header::born_valid(page, selected_txn) {
        return Some(HeaderProblem::Born);
    }
    if level > MAX_LEVEL || expected_level.is_some_and(|expected| expected != level) {
        return Some(HeaderProblem::Level);
    }
    let expected_type = if level == 0 { LEAF_TYPE } else { BRANCH_TYPE };
    if !page_header::kind_valid(page, expected_type, kind as u32) {
        return Some(HeaderProblem::Type);
    }
    let count = page_header::item_count(page);
    let maximum = if level == 0 {
        LEAF_WORDS
    } else {
        BRANCH_CHILDREN
    };
    if count == 0 || count > maximum {
        return Some(HeaderProblem::Header);
    }
    None
}

pub(crate) const fn page_lower(level: u16) -> usize {
    if level == 0 {
        LEAF_END
    } else {
        BRANCH_END
    }
}

pub(crate) fn reserved_zero<S: ByteSource>(page: S, level: u16) -> bool {
    let lower = page_lower(level);
    page.all_zero(lower, PAGE_SIZE - lower)
}

pub(crate) fn initialize<D: PageSink>(page: &mut D, txn: u64, level: u16, kind: Kind) {
    page.fill(0);
    page.write(0, &PAGE_MAGIC)
        .expect("fixed bitmap header fits");
    page.set_byte(
        page_header::TYPE,
        if level == 0 { LEAF_TYPE } else { BRANCH_TYPE },
    )
    .expect("fixed bitmap header fits");
    page.put_u16(page_header::HEADER_BYTES, HEADER_SIZE as u16)
        .expect("fixed bitmap header fits");
    page.put_u64(page_header::BORN_TXN, txn)
        .expect("fixed bitmap header fits");
    page.put_u16(page_header::LEVEL, level)
        .expect("fixed bitmap header fits");
    page.put_u16(page_header::LOWER, page_lower(level) as u16)
        .expect("fixed bitmap header fits");
    page.put_u16(page_header::UPPER, PAGE_SIZE as u16)
        .expect("fixed bitmap header fits");
    page.put_u32(page_header::AUX, kind as u32)
        .expect("fixed bitmap header fits");
}

pub(crate) fn leaf_word<S: ByteSource>(page: S, index: usize) -> Result<u64> {
    if index >= LEAF_WORDS {
        return Err(Error::Corrupt("bitmap word index is invalid"));
    }
    Ok(u64_le(page, HEADER_SIZE + index * 8))
}

pub(crate) fn set_leaf_word<D: PageEdit>(page: &mut D, index: usize, word: u64) -> Result<()> {
    if index >= LEAF_WORDS {
        return Err(Error::Corrupt("bitmap word index is invalid"));
    }
    page.put_u64(HEADER_SIZE + index * 8, word)
}

pub(crate) fn branch_child<S: ByteSource>(page: S, index: usize) -> Result<u32> {
    if index >= BRANCH_CHILDREN {
        return Err(Error::Corrupt("bitmap child index is invalid"));
    }
    Ok(u32_le(page, HEADER_SIZE + 32 + index * 4))
}

pub(crate) fn set_branch_child<D: PageEdit>(page: &mut D, index: usize, child: u32) -> Result<()> {
    if index >= BRANCH_CHILDREN {
        return Err(Error::Corrupt("bitmap child index is invalid"));
    }
    page.put_u32(HEADER_SIZE + 32 + index * 4, child)
}

pub(crate) fn summary_bit<S: ByteSource>(page: S, index: usize) -> Result<bool> {
    if index >= BRANCH_CHILDREN {
        return Err(Error::Corrupt("bitmap summary index is invalid"));
    }
    Ok(u64_le(page, HEADER_SIZE + (index / 64) * 8) & (1u64 << (index % 64)) != 0)
}

pub(crate) fn set_summary<D: PageEdit>(page: &mut D, index: usize, value: bool) -> Result<()> {
    if index >= BRANCH_CHILDREN {
        return Err(Error::Corrupt("bitmap summary index is invalid"));
    }
    let at = HEADER_SIZE + (index / 64) * 8;
    let mask = 1u64 << (index % 64);
    let word = u64_le(page.view(), at);
    page.put_u64(at, if value { word | mask } else { word & !mask })
}

pub(crate) fn first_summary<S: ByteSource>(page: S, start: usize) -> Option<usize> {
    if start >= BRANCH_CHILDREN {
        return None;
    }
    let mut word_index = start / 64;
    let mut word = u64_le(page, HEADER_SIZE + word_index * 8) & (u64::MAX << (start % 64));
    loop {
        if word != 0 {
            return Some(word_index * 64 + word.trailing_zeros() as usize);
        }
        word_index += 1;
        if word_index == 4 {
            return None;
        }
        word = u64_le(page, HEADER_SIZE + word_index * 8);
    }
}

pub(crate) fn first_leaf_word<S: ByteSource>(page: S) -> Option<(usize, u64)> {
    (0..LEAF_WORDS).find_map(|index| {
        let value = leaf_word(page, index).expect("bounded bitmap word index");
        (value != 0).then_some((index, value))
    })
}

pub(crate) fn nonzero_leaf_words<S: ByteSource>(page: S) -> usize {
    (0..LEAF_WORDS)
        .filter(|&index| leaf_word(page, index).expect("bounded bitmap word index") != 0)
        .count()
}

pub(crate) fn coverage(level: u16) -> Result<u64> {
    let mut value = LEAF_BITS;
    for _ in 0..level {
        value = value
            .checked_mul(BRANCH_CHILDREN as u64)
            .ok_or(Error::ArithmeticOverflow("bitmap coverage"))?;
    }
    Ok(value)
}

pub(crate) fn required_level(limit: u64) -> Result<u16> {
    if limit == 0 || limit > 1u64 << 32 {
        return Err(Error::InvalidArgument("bitmap limit is invalid"));
    }
    for level in 0..=MAX_LEVEL {
        if coverage(level)? >= limit {
            return Ok(level);
        }
    }
    Err(Error::ArithmeticOverflow("bitmap limit"))
}

pub(crate) fn leaf_word_index(bit: u32) -> usize {
    ((u64::from(bit) % LEAF_BITS) / 64) as usize
}

pub(crate) fn child_index(bit: u32, level: u16) -> Result<usize> {
    Ok(((u64::from(bit) / coverage(level - 1)?) % BRANCH_CHILDREN as u64) as usize)
}

File: v4/go/internal/bitmap/bitmap_page.go
// Package bitmap implements the canonical hierarchical-bitmap page layout
// and its COW used/free mutation, mirroring the Rust bitmap_page and
// free_bitmap modules. The free bitmap is the allocation core's free-page
// authority; the used bitmap supports the dictionary ID allocators that
// arrive with the edit workflows.
package bitmap

import (
	"bytes"
	"math/bits"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/work"
)

// Layout constants (binary-format-v4.md section 9, Rust bitmap_page.rs).
const (
	LeafWords         = format.BitmapLeafWords // 500
	LeafBits          = uint64(LeafWords * 64) // 32000
	BranchChildren    = format.BitmapFanout    // 256
	leafWordsOffset   = format.SlottedHeaderSize
	branchSummarySize = 32
	BranchChildrenOff = format.SlottedHeaderSize + branchSummarySize
	LeafEnd           = leafWordsOffset + LeafWords*8
	BranchEnd         = BranchChildrenOff + BranchChildren*4
	MaxLevel          = 3
	summaries         = 4 // 4 summary words of 64 bits = 256 bits
)

// Kind discriminates the four bitmap namespaces (Rust Kind).
type Kind uint32

const (
	KindFree       Kind = 1
	KindFeed       Kind = 2
	KindMembership Kind = 3
	KindStructure  Kind = 4
)

// FirstCandidate is the lowest allocatable bit of the kind.
func (k Kind) FirstCandidate() uint64 {
	if k == KindMembership || k == KindStructure {
		return 1
	}
	return 0
}

// Header is the validated geometry of one bitmap page.
type Header struct {
	Level     uint16
	ItemCount int
}

// PageLower is the fixed lower bound of one bitmap page level (Rust
// page_lower).
func PageLower(level uint16) int {
	if level == 0 {
		return LeafEnd
	}
	return BranchEnd
}

// HeaderProblemNone reports a healthy bitmap header; the other classes
// mirror Rust bitmap_page::HeaderProblem and select the validation
// reason.
type HeaderProblem uint8

const (
	HeaderProblemNone HeaderProblem = iota
	HeaderProblemHeader
	HeaderProblemBorn
	HeaderProblemLevel
	HeaderProblemType
)

// CheckedHeader inspects one bitmap page and classifies its defect (Rust
// bitmap_page::checked_header): Header, Born, Level, or Type select the
// validation reason classes, and the geometry is returned only for a
// healthy header.
func CheckedHeader(page []byte, selectedTxn uint64, kind Kind, expectedLevel *uint16) (Header, HeaderProblem) {
	level := format.U16(page[format.HeaderLevel:])
	lower := PageLower(level)
	// Rust page_header::common_valid runs on every bitmap parse (private
	// and committed): magic, zero flags, and the fixed 32-byte header
	// size. A torn write at offset 0 with the rest of the header intact
	// must be rejected here, before any COW copy re-seals the page with
	// a valid checksum. The magic compare is bounded (4 bytes) and never
	// moves mapped bytes into owned memory.
	if len(page) != format.PageSize ||
		!bytes.Equal(page[format.HeaderMagic:format.HeaderMagic+4], format.PageMagic[:]) ||
		page[format.HeaderFlags] != 0 ||
		int(format.U16(page[format.HeaderSizePos:])) != format.SlottedHeaderSize ||
		int(format.U16(page[format.HeaderLower:])) != lower ||
		int(format.U16(page[format.HeaderUpper:])) != format.PageSize {
		return Header{}, HeaderProblemHeader
	}
	born := format.U64(page[format.HeaderBorn:])
	if born == 0 || born > selectedTxn {
		return Header{}, HeaderProblemBorn
	}
	if level > MaxLevel || (expectedLevel != nil && *expectedLevel != level) {
		return Header{}, HeaderProblemLevel
	}
	expectedType := format.PageTypeBitmapLeaf
	if level != 0 {
		expectedType = format.PageTypeBitmapBranch
	}
	if page[format.HeaderType] != byte(expectedType) || format.U32(page[format.HeaderAux:]) != uint32(kind) {
		return Header{}, HeaderProblemType
	}
	count := int(format.U16(page[format.HeaderCount:]))
	maximum := LeafWords
	if level != 0 {
		maximum = BranchChildren
	}
	if count == 0 || count > maximum {
		return Header{}, HeaderProblemHeader
	}
	return Header{Level: level, ItemCount: count}, HeaderProblemNone
}

// InspectHeader validates one bitmap page and returns its geometry (Rust
// bitmap_page::inspect_header); the classified defect becomes the
// Corrupt class for the healthy-read callers.
func InspectHeader(page []byte, selectedTxn uint64, kind Kind, expectedLevel *uint16) (Header, error) {
	header, problem := CheckedHeader(page, selectedTxn, kind, expectedLevel)
	if problem != HeaderProblemNone {
		return Header{}, corrupt(headerProblemDetail(problem))
	}
	return header, nil
}

// headerProblemDetail is the Corrupt message of one header problem
// class (the pre-classification messages are unchanged).
func headerProblemDetail(problem HeaderProblem) string {
	switch problem {
	case HeaderProblemBorn:
		return "bitmap page transaction is invalid"
	case HeaderProblemLevel:
		return "bitmap page level is invalid"
	case HeaderProblemType:
		return "bitmap page type or discriminator is invalid"
	default:
		return "bitmap page header is invalid"
	}
}

// ReservedZero reports whether the reserved tail is all zeroes (Rust
// reserved_zero).
func ReservedZero(page []byte, level uint16) bool {
	lower := PageLower(level)
	for at := lower; at < format.PageSize; at++ {
		if page[at] != 0 {
			return false
		}
	}
	return true
}

// Initialize stamps one fresh bitmap page (Rust bitmap_page::initialize).
func Initialize(page []byte, txn uint64, level uint16, kind Kind) {
	format.InitializePageHeader(page, pageTypeForLevel(level), txn, 0, level, uint16(PageLower(level)), format.PageSize, uint32(kind))
}

func pageTypeForLevel(level uint16) format.PageType {
	if level == 0 {
		return format.PageTypeBitmapLeaf
	}
	return format.PageTypeBitmapBranch
}

// LeafWord reads one 8-byte bitmap word (Rust leaf_word).
func LeafWord(page []byte, index int) (uint64, error) {
	if index >= LeafWords {
		return 0, corrupt("bitmap word index is invalid")
	}
	work.BitmapProbe(1)
	return format.U64(page[leafWordsOffset+index*8:]), nil
}

// SetLeafWord writes one 8-byte bitmap word.
func SetLeafWord(page []byte, index int, word uint64) error {
	if index >= LeafWords {
		return corrupt("bitmap word index is invalid")
	}
	format.PutU64(page[leafWordsOffset+index*8:], word)
	work.BytesMoved(8) // Rust bitmap_page: page.put_u64
	return nil
}

// BranchChild reads one child page number (Rust branch_child).
func BranchChild(page []byte, index int) (uint32, error) {
	if index >= BranchChildren {
		return 0, corrupt("bitmap child index is invalid")
	}
	work.BitmapProbe(1)
	return format.U32(page[BranchChildrenOff+index*4:]), nil
}

// SetBranchChild writes one child page number.
func SetBranchChild(page []byte, index int, child uint32) error {
	if index >= BranchChildren {
		return corrupt("bitmap child index is invalid")
	}
	format.PutU32(page[BranchChildrenOff+index*4:], child)
	work.BytesMoved(4) // Rust bitmap_page: page.put_u32
	return nil
}

// CheckedBranchChild reads and bounds-checks one child (Rust
// checked_branch_child).
func CheckedBranchChild(page []byte, header Header, index int, pageLimit uint64) (uint32, error) {
	if header.Level == 0 || index >= BranchChildren {
		return 0, corrupt("bitmap child lookup is invalid")
	}
	child, err := BranchChild(page, index)
	if err != nil {
		return 0, err
	}
	if child != 0 && (child < 2 || uint64(child) >= pageLimit) {
		return 0, corrupt("bitmap child is outside page bounds")
	}
	return child, nil
}

// ReplaceBranchChild writes one child plus its summary bit and returns the
// new child count (Rust replace_branch_child).
func ReplaceBranchChild(page []byte, header Header, index int, child uint32, summary bool) (int, error) {
	if header.Level == 0 || index >= BranchChildren {
		return 0, corrupt("bitmap child index is invalid")
	}
	old, err := BranchChild(page, index)
	if err != nil {
		return 0, err
	}
	if err := SetBranchChild(page, index, child); err != nil {
		return 0, err
	}
	if err := SetSummary(page, index, summary); err != nil {
		return 0, err
	}
	count := header.ItemCount
	if old == 0 && child != 0 {
		count++
	}
	if old != 0 && child == 0 {
		count--
	}
	if count < 0 {
		return 0, corrupt("bitmap child count underflows")
	}
	format.PutU16(page[format.HeaderCount:], uint16(count))
	work.BytesMoved(2) // Rust set_count: page.put_u16
	return count, nil
}

// SummaryBit reads one branch summary bit (Rust summary_bit).
func SummaryBit(page []byte, index int) (bool, error) {
	if index >= BranchChildren {
		return false, corrupt("bitmap summary index is invalid")
	}
	work.BitmapProbe(1)
	at := format.SlottedHeaderSize + (index/64)*8
	return format.U64(page[at:])&(1<<(index%64)) != 0, nil
}

// SetSummary writes one branch summary bit (Rust set_summary).
func SetSummary(page []byte, index int, value bool) error {
	if index >= BranchChildren {
		return corrupt("bitmap summary index is invalid")
	}
	at := format.SlottedHeaderSize + (index/64)*8
	mask := uint64(1 << (index % 64))
	word := format.U64(page[at:])
	if value {
		word |= mask
	} else {
		word &^= mask
	}
	format.PutU64(page[at:], word)
	work.BytesMoved(8) // Rust summary bit set: page.put_u64
	return nil
}

// FirstSummary returns the first set summary bit at or after start and
// whether one exists (Rust first_summary's Option).
func FirstSummary(page []byte, start int) (int, bool, error) {
	if start >= BranchChildren {
		return 0, false, nil
	}
	wordIndex := start / 64
	word := format.U64(page[format.SlottedHeaderSize+wordIndex*8:]) & (^uint64(0) << (start % 64))
	for {
		if word != 0 {
			return wordIndex*64 + trailingZeros(word), true, nil
		}
		wordIndex++
		if wordIndex == summaries {
			return 0, false, nil
		}
		word = format.U64(page[format.SlottedHeaderSize+wordIndex*8:])
	}
}

func trailingZeros(word uint64) int { return bits.TrailingZeros64(word) }

// FirstLeafWord returns the first nonzero leaf word (Rust first_leaf_word).
func FirstLeafWord(page []byte) (int, uint64, error) {
	for index := 0; index < LeafWords; index++ {
		value, err := LeafWord(page, index)
		if err != nil {
			return 0, 0, err
		}
		if value != 0 {
			return index, value, nil
		}
	}
	return 0, 0, nil
}

// NonzeroChildren counts the nonzero branch children and verifies that
// every summary bit agrees with its child (Rust free_bitmap
// nonzero_children).
func NonzeroChildren(page []byte) (int, error) {
	count := 0
	for index := 0; index < BranchChildren; index++ {
		child, err := BranchChild(page, index)
		if err != nil {
			return 0, err
		}
		summary, err := SummaryBit(page, index)
		if err != nil {
			return 0, err
		}
		if summary != (child != 0) {
			return 0, corrupt("free bitmap summary disagrees with child")
		}
		if child != 0 {
			count++
		}
	}
	return count, nil
}

// NonzeroLeafWords counts the nonzero leaf words (Rust nonzero_leaf_words).
func NonzeroLeafWords(page []byte) (int, error) {
	count := 0
	for index := 0; index < LeafWords; index++ {
		value, err := LeafWord(page, index)
		if err != nil {
			return 0, err
		}
		if value != 0 {
			count++
		}
	}
	return count, nil
}

// Coverage returns the bit span of one level (Rust coverage).
func Coverage(level uint16) (uint64, error) {
	value := LeafBits
	for at := 0; at < int(level); at++ {
		value *= BranchChildren
	}
	return value, nil
}

// RequiredLevel returns the smallest level covering limit bits (Rust
// required_level).
func RequiredLevel(limit uint64) (uint16, error) {
	if limit == 0 || limit > 1<<32 {
		return 0, invalid("bitmap limit is invalid")
	}
	for level := uint16(0); level <= MaxLevel; level++ {
		coverage, err := Coverage(level)
		if err != nil {
			return 0, err
		}
		if coverage >= limit {
			return level, nil
		}
	}
	return 0, corrupt("bitmap limit")
}

// LeafWordIndex returns the word holding bit (Rust leaf_word_index).
func LeafWordIndex(bit uint32) int {
	return int((uint64(bit) % LeafBits) / 64)
}

// ChildIndex returns the branch child slot of bit at level (Rust
// child_index).
func ChildIndex(bit uint32, level uint16) (int, error) {
	coverage, err := Coverage(level - 1)
	if err != nil {
		return 0, err
	}
	return int((uint64(bit) / coverage) % BranchChildren), nil
}

func corrupt(detail string) error {
	return &format.Error{Code: format.CodeFormatInvalid, Detail: detail}
}

func invalid(detail string) error {
	return &format.Error{Code: format.CodeInvalidArgument, Detail: detail}
}

func overflow(detail string) error {
	return &format.Error{Code: format.CodeArithmeticOverflow, Detail: detail}
}
